lib/movable_ink/aws/metadata.rb

```diff
@@ -0,0 +1,62 @@
+require 'net/http'
+
+module MovableInk
+  class AWS
+    module Metadata
+      def http
+        @http ||= begin
+          http = Net::HTTP.new("169.254.169.254", 80)
+          http.open_timeout = 1
+          http.read_timeout = 1
+          http
+        end
+      end
+
+      def retrieve_metadata(key)
+        request = Net::HTTP::Get.new("/latest/meta-data/#{key}")
+        request['X-aws-ec2-metadata-token'] = imds_token
+        response = http.request(request)
+        response.body
+      rescue
+        ""
+      end
+
+      def availability_zone
+        @availability_zone ||= begin
+          az = retrieve_metadata('placement/availability-zone')
+          raise(MovableInk::AWS::Errors::EC2Required) if az.empty?
+          az
+        end
+      end
+
+      def instance_id
+        @instance_id ||= begin
+          id = retrieve_metadata('instance-id')
+          raise(MovableInk::AWS::Errors::EC2Required) if id.empty?
+          id
+        end
+      end
+
+      def private_ipv4
+        @ipv4 ||= begin
+          ipv4 = retrieve_metadata('local-ipv4')
+          raise(MovableInk::AWS::Errors::EC2Required) if ipv4.empty?
+          ipv4
+        end
+      end
+
+      private
+
+      def imds_token
+        @token ||= begin
+          request = Net::HTTP::Put.new('/latest/api/token')
+          request['X-aws-ec2-metadata-token-ttl-seconds'] = 120
```

Comment:
The API supports tokens that are valid for up to 6 hours (21600s).  2 minutes (120s) feels like overkill here, but I'm erring on the side of simplicity avoiding the need for any kind of token refresh logic.  I can't think of anything we use that keeps an instance of the gem around for longer than a few seconds max.  Maybe the rails app?

Reply:
As I think about this more, the AWS documentation at https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html says:
```
You can choose to reuse a token or create a new token with every request. For a small number of requests, it might be easier to generate and immediately use a token each time you need to access the instance metadata service.
```
Given that we only use the metadata service to retrieve three pieces of data (instance ID, private IP, availability zone), and we memoize those three pieces of data to avoid retrieving them multiple times (they don't change during the life of an instance), maybe it makes sense not to memoize the token and retrieve it every time we call the metadata service.  Thoughts?

Reply:
Yeah, for what we're doing it seems like it makes sense to not memoize the token.

Reply:
The deploy process and scripts that provisioning uses would likely keep the gem around for much longer than 2 minutes. We will probably need a longer timeout or need to refresh the token.

Reply:
I've gotten rid of the memoization of the token.  I think it's fine to make three calls for the token to retrieve the instance ID, IP, and AZ.